### PR TITLE
CYS - Move the `ai/store-title` endpoint to woocommerce admin api

### PR DESCRIPTION
**Pull request by albarin**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Part of https://github.com/woocommerce/woocommerce/issues/48150

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. In a `WooExpress` store, go to `wp-admin/admin.php?page=wc-admin&path=%2Fcustomize-store`.
2. Start designing a new store and write the description.
3. Confirm the flow finishes without errors and you see a store name related to your store description in the assembler.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
CYS - Move the ai/store-title endpoint to woocommerce admin API
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Admin/API/AI/Middleware.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+use Automattic\WooCommerce\StoreApi\Exceptions\RouteException;
+use WP_Error;
+
+/**
+ * Middleware class.
+ *
+ * @internal
+ */
+class Middleware {
```
Comment (albarin) [inline plugins/woocommerce/src/Admin/API/AI/Middleware.php:15]: This is just a copy of this class https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/StoreApi/Routes/V1/AI/Middleware.php, the original one will be removed once all endpoints and migrated.

**Comment by tjcafferkey** [comment]
I don't think its related to this PR but is it common for errors like this to happen?

Screenshot 2024-08-07 at 09 02 24

**Comment by tjcafferkey** [comment]
>a store name related to your store description in the assembler.

When you say a store name do you mean this setting? Because nothing changed for me

Screenshot 2024-08-07 at 09 28 36

Screenshot 2024-08-07 at 09 28 13

**Comment by albarin** [comment]
> When you say a store name do you mean this setting? Because nothing changed for me


@tjcafferkey Yes, the title will only be AI-generated when it's [empty or the default](https://github.com/woocommerce/woocommerce/blob/e20641341de53901d7bef1d5dffbe980e611898a/plugins/woocommerce/src/StoreApi/Routes/V1/AI/StoreTitle.php#L106) (`Site Title`), so in this case it's normal it wasn't updated. Could you remove or update it to the default and re-run the flow?

**Comment by tjcafferkey** [comment]
It worked for an empty site title but not the default "Site Title"

Screenshot 2024-08-07 at 09 37 36
Screenshot 2024-08-07 at 09 38 47

**Comment by tjcafferkey** [comment]
Maybe we can normalize the site title values?

`strtolower( trim( self::DEFAULT_TITLE ) ) === strtolower( trim( $store_title ) )`

File: plugins/woocommerce/src/Admin/API/AI/StoreTitle.php [hunk 2]
```diff
@@ -40,63 +30,58 @@ class StoreTitle extends AbstractRoute {
...
-			return $this->error_to_response(
-				new \WP_Error(
-					'invalid_business_description',
-					__( 'Invalid business description.', 'woocommerce' )
-				)
+			return new WP_Error(
+				'invalid_business_description',
+				__( 'Invalid business description.', 'woocommerce' )
 			);
 		}
```
Comment (tjcafferkey) [inline plugins/woocommerce/src/Admin/API/AI/StoreTitle.php:88]: Can we update `$store_title = html_entity_decode( get_option( 'blogname' ) );` to use the `STORE_TITLE_OPTION_NAME` const

File: plugins/woocommerce/src/Admin/API/AI/StoreTitle.php [hunk 3]
```diff
@@ -40,76 +32,63 @@ class StoreTitle extends AbstractRoute {
...
 
-		$store_title                 = html_entity_decode( get_option( 'blogname' ) );
+		$store_title                 = html_entity_decode( get_option( self::STORE_TITLE_OPTION_NAME ) );
 		$previous_ai_generated_title = html_entity_decode( get_option( 'ai_generated_site_title' ) );
```
Comment (tjcafferkey) [inline plugins/woocommerce/src/Admin/API/AI/StoreTitle.php]: Non-blocking nitpick: If we're storing `blogname` in a class variable I feel like we should do the same with `ai_generated_site_title`

**Comment by tjcafferkey** [comment]
I just got the following error from WP.com

```
Error Details
=============
An error of type E_ERROR was caused in line 19 of the file /srv/htdocs/wp-content/plugins/woocommerce/src/Admin/API/AI/StoreTitle.php. Error message: Uncaught Error: Class "Automattic\WooCommerce\Admin\API\AI\AIEndpoint" not found in /srv/htdocs/wp-content/plugins/woocommerce/src/Admin/API/AI/StoreTitle.php:19
Stack trace:
#0 /wordpress/plugins/jetpack/13.7-beta/vendor/jetpack-autoloader/class-php-autoloader.php(90): require()
#1 /srv/htdocs/wp-content/plugins/woocommerce/src/Admin/API/Init.php(164): Automattic\Jetpack\Autoloader\jpf11009ded9fc4592b6a05b61ce272b3c_jetpackⓥ13_7_beta\al3_0_9\PHP_Autoloader::load_class('Automattic\\WooC...')
#2 /wordpress/core/6.6.1/wp-includes/class-wp-hook.php(324): Automattic\WooCommerce\Admin\API\Init->rest_api_init(Object(WP_REST_Server))
#3 /wordpress/core/6.6.1/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters(NULL, Array)
#4 /wordpress/core/6.6.1/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#5 /wordpress/core/6.6.1/wp-includes/rest-api.php(589): do_action('rest_api_init', Object(WP_REST_Server))
#6 /wordpress/core/6.6.1/wp-includes/rest-api.php(547): rest_get_server()
#7 /wordpress/core/6.6.1/wp-includes/rest-api.php(2925): rest_do_request(Object(WP_REST_Request))
#8 [internal function]: rest_preload_api_request(Array, '/wc-analytics/r...')
#9 /srv/htdocs/wp-content/plugins/woocommerce/src/Internal/Admin/Settings.php(145): array_reduce(Array, 'rest_preload_ap...')
#10 /wordpress/core/6.6.1/wp-includes/class-wp-hook.php(324): Automattic\WooCommerce\Internal\Admin\Settings->add_component_settings(Array)
#11 /wordpress/core/6.6.1/wp-includes/plugin.php(205): WP_Hook->apply_filters(Array, Array)
#12 /srv/htdocs/wp-content/plugins/woocommerce/src/Internal/Admin/WCAdminSharedSettings.php(61): apply_filters('woocommerce_adm...', Array)
#13 /srv/htdocs/wp-content/plugins/woocommerce/src/Blocks/Assets/AssetDataRegistry.php(271): Automattic\WooCommerce\Internal\Admin\WCAdminSharedSettings->Automattic\WooCommerce\Internal\Admin\{closure}()
#14 /srv/htdocs/wp-content/plugins/woocommerce/src/Blocks/Assets/AssetDataRegistry.php(387): Automattic\WooCommerce\Blocks\Assets\AssetDataRegistry->execute_lazy_data()
#15 /wordpress/core/6.6.1/wp-includes/class-wp-hook.php(324): Automattic\WooCommerce\Blocks\Assets\AssetDataRegistry->enqueue_asset_data('')
#16 /wordpress/core/6.6.1/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters('', Array)
#17 /wordpress/core/6.6.1/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#18 /wordpress/core/6.6.1/wp-admin/admin-footer.php(95): do_action('admin_print_foo...')
#19 /wordpress/core/6.6.1/wp-admin/update.php(186): require_once('/wordpress/core...')
#20 {main}
 thrown
```

**Comment by albarin** [comment]
> I just got the following error from WP.com

@tjcafferkey I think you need to run `pnpm i`.

File: plugins/woocommerce/src/Admin/API/AI/StoreTitle.php (hunk 3 above)
Comment (albarin) [reply plugins/woocommerce/src/Admin/API/AI/StoreTitle.php]: 👍

**Comment by tjcafferkey** [comment]
This seems to work fine apart from when "Site Title" is default, it doesn't update with an AI generated one

**Review by tjcafferkey — APPROVED** [review]

**Pull request merged**